camera_models.py

```diff
@@ -1129,4 +1130,171 @@ def get_rays_wrt_lidar_frame(self):
+        '''
+        
+        pixel_coor = self.in_wrap(pixel_coor) # Bx2xN
+        
+        # mx and my becomes float64 if pixel_coor.dtype is integer type.
+        # Convert pixel to angle on the unit sphere.
+
+        mx = ( pixel_coor[..., 0, :] - self.ss.W / 2)
+        my = ( pixel_coor[..., 1, :] - self.ss.H / 2)
+        r = (mx**2.0 + my**2.0)**0.5
+
+        # Compute theta and phi. Theta is the angle of 3D points from the z-axis. Phi is the angle of 3D points from the x-axis on the x-y plane.
+        th = torch.atan2( my, mx )
+        ph = r * self.fov_rad / self.ss.W
+
+        # Valid rays are only those within the field of view.
+        valid_mask = torch.abs(ph) <= self.fov_rad / 2.0
+
+        # Compute the 3D points. Those are of unit length.
+        x = torch.sin(ph) * torch.cos(th)
+        y = torch.sin(ph) * torch.sin(th)
+        z = torch.cos(ph)
+
+        # Need to deal with batch dim
+        ray = torch.stack( (x, y, z), dim=-2 )
+
+        return self.out_wrap(ray), self.out_wrap(valid_mask)
+
+    def point_3d_2_pixel(self, point_3d, normalized=False):
+        '''
+        Arguments:
+        point_3d (Tensor): A 3xN Tensor contains 3D point coordinates. 
+        normalized (bool): If True, then the returned coordinates are normalized to [-1, 1]
+        
+        NOTE: point_3d can also have a dimension of Bx3xN, where B is the 
+        batch number. 
+        
+        Returns: 
+        pixel_coor: A 2xN Tensor representing the 2D pixels. Bx2xN if batched.
+        valid_mask: A (N,) Tensor representing the valid mask. BXN if batched.
+        '''
+
+        point_3d = self.in_wrap(point_3d)
+
+        # torch.split results in Bx1XN.
+        x, y, z = torch.split( point_3d, 1, dim=-2 )
+
+        x2 = x**2.0 # Note: this may promote x2 to torch.float64 if point_3d.dtype=torch.int. 
+        y2 = y**2.0
+        z2 = z**2.0
+
+        # Magnitude of the 3D point.
+        R = (x2 + y2 + z2)**0.5
+
+        # Theta (th) is the angle between the ray and the z-axis.
+        # Phi (ph) is the angle between the ray and the x-axis, in the x-y plane.
+        ph = torch.acos( z / R )
+        th = torch.atan2( y, x )
+
+
+        # The radius in the image plane.
+        r = ph * self.ss.W / self.fov_rad
+
+        # The pixel coordinates.
+        # If any are FTensors, then convert to Tensor.
+        if isinstance(th, FTensor):
+            th = th.tensor()
+            ph = ph.tensor()
+
+
+        u = r * torch.cos(th) + self.ss.W / 2
+        v = r * torch.sin(th) + self.ss.H / 2
+
+        # Mask out the invalid pixels.
+        valid_mask = (u >= 0) & (u < self.ss.W) & (v >= 0) & (v < self.ss.H)
+        
+        # Pixel coordinates.
+        if normalized:
+            u = u / self.ss.W * 2 - 1
+            v = v / self.ss.H * 2 - 1
+
+        pixel_coor = torch.cat( (u, v), dim=-2 )
+
+        # Filter by FOV.
+        a = x2y2z_2_z_angle( x2, y2, z )

```

Comment:
Filter by FOV may not be needed?

Reply:
Why not?
What I am thinking is that without it points behind the camera might still be allowed to be projected onto the frame.

Reply:
On Line 1231 of the current file
```python
r = ph * self.ss.W / self.fov_rad
```
This asures that the border of the image is at `self.fov_rad / 2 angle`. And Line 1244
```python
valid_mask = (u >= 0) & (u < self.ss.W) & (v >= 0) & (v < self.ss.H)
```
makes sure that anything beyond the image border (larger than the FoV) will be masked.

Reply:
Great! This is what I was missing. Will take it out.